src/utils/api.ts

```diff
@@ -74,5 +75,133 @@ export async function verifyPlatformUrl(config: Config): Promise<{ success: bool
 
 export function initEndpoint(config: Config): MarkupAIClient {
   const platformUrl = getPlatformUrl(config);
-  return new MarkupAIClient({ token: config.apiKey, baseUrl: platformUrl });
+  const rawClient = new MarkupAIClient({ token: config.apiKey, baseUrl: platformUrl });
+  return wrapClientWithRateLimit(rawClient, config);
+}
+
+// Generic rate limit aware retry helper for SDK calls
+export async function withRateLimitRetry<T>(
+  operation: () => Promise<T>,
+  config: Config,
+  operationLabel?: string,
+): Promise<T> {
+  const maxRetries = config.rateLimit?.maxRetries ?? 5;
+  const baseDelay = config.rateLimit?.initialDelayMs ?? 1000;
+  const maxDelay = config.rateLimit?.maxDelayMs ?? 16000;
+  const jitter = config.rateLimit?.jitter ?? true;
+
+  let attempt = 0;
+  for (;;) {
+    try {
+      return await operation();
+    } catch (err) {
+      type RateLimitLike = {
+        statusCode?: number;
+        status?: number;
+        body?: Record<string, unknown> | unknown;
+        headers?: Record<string, unknown>;
+      };
+      const error = err as RateLimitLike | Error;
+
+      const status = (error as RateLimitLike)?.statusCode ?? (error as RateLimitLike)?.status ?? undefined;
+      const isRateLimit = status === 429;
+
+      if (!isRateLimit) {
+        // rethrow non-rate limit errors immediately
+        throw err;
+      }
+
+      if (attempt >= maxRetries) {
+        // Convert into our ApiError with rate limit type
+        if ('statusCode' in (error as RateLimitLike)) {
+          throw ApiError.fromResponse(
+            429,
+            ((error as RateLimitLike).body as Record<string, unknown>) ?? { detail: 'Rate limit exceeded' },
+          );
+        }
+        throw new ApiError('Rate limit exceeded', ErrorType.RATE_LIMIT_ERROR, 429, {});
```

Comment:
I'm wondering if original error should be added as cause instead of swallowing.

Reply:
You're right. I've updated the rate-limit retry to preserve the original error as the cause instead of swallowing it.